File: Python 3/Problem_019.py
# How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?


WEEK = {"SUN": 0, "MON": 1, "TUE": 2, "WED": 3, "THU": 4, "FRI": 5, "SAT": 6}
YEAR = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
LEAP = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]


def is_leap(year):
    if year % 400 == 0:
        return True
    if year % 100 == 0:
        return False
    if year % 4 == 0:
        return True
    return False


def advance_time(today, days):
    return (today + days) % len(WEEK)


def count_day_in_20_century(day_to_count, start_year=1901, end_year=2000):
    assert 1900 <= start_year <= end_year, "Year is not 20th century"
    BEGINNING_OF_1900 = WEEK["MON"]

    day = BEGINNING_OF_1900
    for y in range(1900, start_year):
        print(y)
        year_length = 365 if is_leap(y) else 366
        day = advance_time(day, year_length)

    count = 0
    for year in range(start_year, end_year + 1):
        if is_leap(year):
            for month in LEAP:
                day = advance_time(day, month)
                count += day == WEEK[day_to_count]
        else:
            for month in YEAR:
                day = advance_time(day, month)
                count += day == WEEK[day_to_count]

    return count


if __name__ == "__main__":
    print(count_day_in_20_century("SUN"))

File: Haskell/Problem_019.hs
-- How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?
import           Data.List                      ( foldl' )
import qualified Data.Time                     as Time


data DayOfWeek = Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday deriving (Eq, Ord, Enum, Show)
data Month = Jan | Feb | Mar | Apr | May | Jun | Jul | Aug | Sep | Oct | Nov | Dec deriving (Eq, Ord, Enum, Show)
type Year = Int


year :: [(Month, Int)]
year = zip [Jan .. Dec] [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]


leapYear :: [(Month, Int)]
leapYear = take 1 year ++ [(Feb, 29)] ++ drop 2 year


isLeap :: Year -> Bool
isLeap y | y `mod` 400 == 0 = True
         | y `mod` 100 == 0 = False
         | y `mod` 4 == 0   = True
         | otherwise        = False


monthsOfYear :: Year -> [(Month, Int)]
monthsOfYear y | isLeap y  = leapYear
               | otherwise = year


week :: [DayOfWeek]
week = [Sunday .. Saturday]


advanceTime :: DayOfWeek -> Int -> DayOfWeek
advanceTime today n_days = future !! max 0 n_days
 where
  future     = restOfWeek ++ cycle week
  restOfWeek = dropWhile (today /=) week


epochTime :: (Year, DayOfWeek)
epochTime = (1900, Monday)
epochYear :: Year
epochYear = fst epochTime
epochDOW :: DayOfWeek
epochDOW = snd epochTime


countDays :: DayOfWeek -> Year -> Year -> Maybe Int
countDays day startY endY | startY < epochYear = Nothing
                          | startY > endY      = Nothing
                          | otherwise          = Just nDays
 where
  nDays       = length $ filter (== day) daysInRange
  daysInRange = scanl advanceAMonth dayAtStartOfRange monthsInRange   where
    advanceAMonth d m = advanceTime d (snd m)
    monthsInRange = concatMap monthsOfYear [startY .. endY]
  dayAtStartOfRange = foldl' advanceAYear epochDOW gapToStart   where
    gapToStart = [epochYear .. startY - 1]
    advanceAYear d y = advanceTime d $ daysInYear y
    daysInYear y = sum [ snd m | m <- monthsOfYear y ]


-- *** *** Doing it with Data.Time *** ***

countDays' :: Time.DayOfWeek -> Int -> Int -> Int
countDays' day startY endY =
  length . filter isThatDay . filter isFirstOfMonth $ daysInRange where
  isThatDay      = (day ==) . Time.dayOfWeek
  isFirstOfMonth = (1 ==) . getDays . Time.toGregorian
    where getDays (_, _, d) = d
  daysInRange = [firstDay .. lastDay]   where -- Time.Day implements Ord, so I can create a range
    firstDay = read $ show startY ++ "-01-01"
    lastDay  = read $ show endY ++ "-12-31"


main :: IO ()
main = do
  print $ countDays Sunday 1901 2000
  print $ countDays' Time.Sunday 1901 2000
